queue/abstract/queue_session.lua

```diff
@@ -0,0 +1,192 @@
+local num_type = qc.num_type
+local str_type = qc.str_type
+
+
+local queue_session = {
+    settings = {}
+}
+
+-- Sessions that don't have any active connections.
+local inactive_sessions = {}
+
+--- Create everything that's needed to work with "shared" sessions.
+local function identification_init()
+    local queue_session_ids = box.space._queue_session_ids
+    if queue_session_ids == nil then
+        queue_session_ids = box.schema.create_space('_queue_session_ids', {
+            temporary = true,
+            format = {
+                {name = 'box_session_id', type = num_type()},
+                {name = 'session_uuid', type = str_type()}
+            }
+        })
+
+        queue_session_ids:create_index('box_sid', {
+            type = 'tree',
+            parts = {1, num_type()},
+            unique = true
+        })
+        queue_session_ids:create_index('uuid', {
+            type = 'tree',
+            parts = {2, str_type()},
+            unique = false
+        })
+    end
+end
+
+--- Get session UUID by connection ID.
+local function get_uuid(connection_id)
+    local session_ids = box.space._queue_session_ids:get(connection_id)
+    local session_uuid
+    if session_ids ~= nil then
+        session_uuid = session_ids[2]
+    else
+        session_uuid = queue_session.identify(connection_id)
+    end
+
+    return session_uuid
+end
+
+local function cleanup_inactive_sessions()
+local cur_time = util.time()
+    for session_uuid, exp_time in pairs(inactive_sessions) do
+        if cur_time >= exp_time then
+            if queue_session._on_session_remove ~= nil then
+                queue_session._on_session_remove(session_uuid)
+            end
+            inactive_sessions[session_uuid] = nil
+        end
+    end
+end
+
+--- Create an expiration fiber to cleanup expired sessions.
+local function create_expiration_fiber()
+    local exp_fiber = fiber.create(function()
+        while true do
+            if box.info.ro == false then
+                local status, err = pcall(cleanup_inactive_sessions)
+                if status == false then
+                    log.error('An error occured while cleanup the sessions: %s',
+                        err)
+                end
+            end
+            fiber.sleep(1)
+        end
+    end)
+
+    return exp_fiber
+end
+
+--- Identify the session.
+-- If session_uuid == nil - create a new UUID for the connection
+-- and return it, otherwise connect to the existing one.
+local function identify(conn_id, session_uuid)
+    local queue_session_ids = box.space._queue_session_ids
+    local session_ids = queue_session_ids.index.box_sid:get(conn_id)
+    local cur_uuid = session_ids and session_ids[2]
+
+    if session_uuid == nil and cur_uuid == nil then
+        -- Generate new UUID for the session.
+        cur_uuid = uuid.bin()
+        queue_session_ids:insert{conn_id, cur_uuid}
+    elseif session_uuid ~= nil then
+        --Validate UUID.
```

Comment:
Usually we put a space between `--` and text.

Reply:
Updated.